lsps2/intercept_handler.go

```diff
@@ -0,0 +1,732 @@
+	isFinal          bool
+	timoutChanClosed bool
+	timeoutChan      chan struct{}
+}
+
+func (p *paymentState) closeTimeoutChan() {
+	if p.timoutChanClosed {
+		return
+	}
+
+	close(p.timeoutChan)
+	p.timoutChanClosed = true
+}
+
+type partState struct {
+	isProcessed bool
+	isFinalized bool
+	req         *shared.InterceptRequest
+	resolution  chan *shared.InterceptResult
+}
+
+type registrationReadyEvent struct {
+	paymentId    string
+	registration *BuyRegistration
+}
+
+type paymentChanOpenedEvent struct {
+	paymentId       string
+	scid            lightning.ShortChannelID
+	channelPoint    *wire.OutPoint
+	htlcMinimumMsat uint64
+}
+
+type paymentFailureEvent struct {
+	paymentId string
+	code      shared.InterceptFailureCode
+}
+
+type awaitingRegistrationEvent struct {
+	paymentId string
+	partId    string
+}
+
+func (i *Interceptor) Start(ctx context.Context) {
+	// Main event loop for stages of htlcs to be handled. Note that the event
+	// loop has to execute quickly, so any code running in the 'handle' methods
+	// must execute quickly. If there is i/o involved, or any waiting, run that
+	// code in a goroutine, and place an event onto the event loop to continue
+	// processing after the slow operation is done.
+	// The nice thing about an event loop is that it runs on a single thread.
+	// So there's no locking needed, as long as everything that needs
+	// synchronization goes through the event loop.
+	for {
+		select {
+		case part := <-i.newPart:
+			i.handleNewPart(part)
+		case ev := <-i.registrationReady:
+			i.handleRegistrationReady(ev)
+		case paymentId := <-i.notRegistered:
+			i.handleNotRegistered(paymentId)
+		case ev := <-i.partAwaitingRegistration:
+			i.handlePartAwaitingRegistration(ev)
+		case paymentId := <-i.paymentReady:
+			i.handlePaymentReady(paymentId)
+		case paymentId := <-i.paymentTimeout:
+			i.handlePaymentTimeout(paymentId)
+		case ev := <-i.paymentFailure:
+			i.handlePaymentFailure(ev.paymentId, ev.code)
+		case ev := <-i.paymentChanOpened:
+			i.handlePaymentChanOpened(ev)
+		}
+	}
+}
+
+func (i *Interceptor) handleNewPart(part *partState) {
+	payment, paymentExisted := i.inflightPayments[part.req.PaymentId()]
+	if !paymentExisted {
+		payment = &paymentState{
+			id:          part.req.PaymentId(),
+			fakeScid:    part.req.Scid,
+			parts:       make(map[string]*partState),
+			timeoutChan: make(chan struct{}),
+		}
+		i.inflightPayments[part.req.PaymentId()] = payment
+
+		go func() {
+			select {
+			case <-time.After(i.config.MppTimeout):
+				// Handle timeout inside the event loop, to make sure there are
+				// no race conditions, since this timeout watcher is running in
+				// a goroutine.
+				i.paymentTimeout <- part.req.PaymentId()
+			case <-payment.timeoutChan:
+				// Stop listening for timeouts when the payment is ready.
+			}
+		}()
+
+		// Fetch the buy registration in a goroutine, to avoid blocking the
+		// event loop.
+		go i.fetchRegistration(part.req.PaymentId(), part.req.Scid)
```

Comment:
The tradeoff of using a go routine here is make the workflow more complicated. I think if we don't use a go routine here then we won't need  these three channels: `awaitingRegistration`  `registrationReady` `notRegistered` which will make the flow more readable and in context.
The disadvantage of having a database query (very fast one) looks negligible to me.

Reply:
I did code it that way at first. The change to put the database query in a goroutine indeed added those chans you mention. The problem is, if the event loop is clugged, that's a fatal issue that's very hard to debug. This event loop processes all htlcs on the node, not just the ones meant for channel creation. If new htlcs keep being added, and the database query does turn out to be slow for some reason, that causes the loop to keep adding new htlcs. And potentially not handle resolutions for channel opening, or process them very slowly. Which can lead to timeouts on the sender (weird behavior), cause us to open unused channels (costly), and in the case of an attack may lead to loss of funds due to underlying htlc timeouts (maybe cln has a mechanism to prevent this, not sure).

So even though it's complicated, I think it's warrranted for the adversarial scenario tbh.

Reply:
That's a good point. I will take a look again to see if we can simplify without risking the regular flow.

Reply:
I understand now the logic behind the decision to choose the go routine for the db query but practically I don't think we had such issue with the db and the simplicity in one channel seems like a great advantage but we can still sleep on it before doing any change. Maybe @yaslama can share some insights.

Reply:
We worked with locks before, based on the payment hash, which means a lot of work can be done in parallel. A slow db operation doesn't affect the general flow then. I added a test here https://github.com/breez/lspd/blob/c6a5798bbd1c6e037c0ee2d636aeb770177c4717/lsps2/intercept_test.go#L707-L760
consisting of 100 payments of 10 parts. Db delay is set to 500ms. That takes around 50 seconds to complete in total without the goroutine. And around 500ms to complete with the goroutine.

Reply:
It would be the same for 100 payments of 1 part. That would also take 50 seconds. with a 500ms delay for fetching the `fetchRegistration` call.

Reply:
I am pretty sure that this indexed query will be executed in no more than 5 millisecond. Most probably even that no io is needed at the db level because it will hit the db in memory cache.
I am not arguing about the performance which I agree your approach is faster, only about the tradeoffs 
Did you consider still using ho routines for parallel ecevution but one go routine for the whole processing if the htlc? So effectively use one channel?

Reply:
It's that you have to wait for other htlcs to be processed in order to continue the current htlc that made me come up with this design. It won't really become less complicated if every htlc is processed in its own goroutine. Now the complexity is now in htlc processing stages on the event loop. If you process them in parallel, the complexity will be in handling locks and races. What I like about this event loop is you don't have to worry about races, because every stage is handled one by one. Everywhere where you need synchronization, like updating the payment state or part state, or finalizing the htlcs for example, you can be certain there is no other thread modifying the same values.

Reply:
BTW you do mention 'indexed query'. I agree things should be fast if everything is indexed correctly. But if there's network maintenance or something, or a query is not indexed properly, or there's even some read lock on the db somewhere that runs for a while, it's problematic in this design. We could try another design, with a single goroutine per htlc, and using locks and waitgroups. That would take away that weird handling of the htlc in stages. I'm not sure whether it would be less complicated though.

Reply:
I see your point. Keep in mind that network maintenance or db not responding in the current design will  keep spinning new go routings that hang.
Let's discuss tomorrow. I think we have all we need to agree on the approach!